src/websock/native/integrate_test.go

```diff
@@ -0,0 +1,508 @@
+// +build !without_amqp
+
+package native_test
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"fmt"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/rinq/httpd/src/websock/native"
+	"github.com/rinq/httpd/src/websock/native/message"
+	"github.com/rinq/rinq-go/src/rinq"
+	"github.com/rinq/rinq-go/src/rinq/options"
+	"github.com/rinq/rinq-go/src/rinqamqp"
+	"github.com/satori/go.uuid"
+	"io"
+	"log"
+	"net/http"
+	"net/http/httptest"
+	"time"
+)
+
+// absolute hack to get around nil being a valid value
+var noBody *struct{ Foo string } = nil
+
+var _ = Describe("handler", func() {
+
+	var (
+		peer rinq.Peer
+
+		req *http.Request
+	)
+
+	const (
+		createSession uint16 = 'S'<<8 | 'C'
+
+		callSync        uint16 = 'C'<<8 | 'C'
+		callSyncSuccess uint16 = 'C'<<8 | 'S'
+		callSyncFailure uint16 = 'C'<<8 | 'F'
+		callSyncError   uint16 = 'C'<<8 | 'E'
+
+		callAsync        uint16 = 'A'<<8 | 'C'
+		callAsyncSuccess uint16 = 'A'<<8 | 'S'
+		callAsyncFailure uint16 = 'A'<<8 | 'F'
+		callAsyncError   uint16 = 'A'<<8 | 'E'
+
+		callExec uint16 = 'C'<<8 | 'X'
+
+		session uint16 = 0xCAFE
+	)
+
+	BeforeEach(func() {
+		req = httptest.NewRequest("GET", "/", nil)
+
+		var err error
+		peer, err = rinqamqp.DialEnv(options.Logger(rinq.NewLogger(false)))
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	Context("integration testing the default settings", func() {
+
+		const (
+			nsBase      = "name-space"
+			cmd         = "cmd"
+			seq    uint = 1
+
+			respBody = "pong"
+		)
+
+		var (
+			ns string
+
+			websocket *mockWebsock
+
+			start chan struct{}
+			kill  chan struct{}
+		)
+
+		BeforeEach(func() {
+
+			ns = nsBase + uuid.NewV4().String()
+
+			start = make(chan struct{})
+			kill = make(chan struct{})
+
+			websocket = &mockWebsock{
+				start: start,
+				dead:  kill,
+				wIn:   make(chan []byte),
+			}
+
+			go func() {
+				defer GinkgoRecover()
+
+				<-start
+				handler := native.NewHandler(peer, message.JSONEncoding)
+				err := handler.Handle(websocket, req)
+				log.Println("got", err.Error(), ", handler closed")
+			}()
+		})
+
+		AfterEach(func() {
+			close(kill)
+			peer.Stop()
+			<-peer.Done()
+		})
+
+		Context("successful calls", func() {
+			BeforeEach(func() {
+				peer.Listen(ns, func(ctx context.Context, req rinq.Request, res rinq.Response) {
+					defer req.Payload.Close()
+					res.Done(rinq.NewPayload(respBody))
+				})
+			})
+
+			It("Sends a successful basic sync ping-pong message", func() {
+				websocket.queueMsg(createSession, session, nil, nil)
```

Comment:
Consider revising the naming of `queueMsg()` and `getMsg()`, as the directionality wasn't clear to me at first.

`websocket` here is representing both the client and the server's view of the connection, so perhaps naming such as: `sendFromClient` and `recvFromServer` or something like that would be more appropriate.

Reply:
The same is probably try for the `wIn` and `rOut` members, since they are used outside the struct.